src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,765 @@
+    if (initDev)
+      {
+        for (int i = 0; i < N; ++i) rgInitialStds[i] = stdDivs[i];
+      }
+      else
+      {
+        for (int i = 0; i < N; ++i) rgInitialStds[i] = double(0.3);
+      }
+
+
+    if (lambda < 2)
+      lambda = 4 + (int) (3.0*log((double) N));
+    if (mu <= 0)
+      mu = lambda / 2;
+
+      weights.set_size(mu);
+      switch(weightMode)
+      {
+      case LINEAR_WEIGHTS:
+        for (int i = 0; i < mu; ++i) weights[i] = mu - i;
+        break;
+      case EQUAL_WEIGHTS:
+        for (int i = 0; i < mu; ++i) weights[i] = 1;
+        break;
+      case LOG_WEIGHTS:
+      default:
+        for (int i = 0; i < mu; ++i) weights[i] = log(mu + 1.) - log(i + 1.);
+        break;
+      }
+
+      // normalize weights vector and set mueff
+      double s1 = 0, s2 = 0;
+      for (int i = 0; i < mu; ++i)
+      {
+        s1 += weights[i];
+        s2 += weights[i]*weights[i];
+      }
+      mueff = s1*s1/s2;
+      for (int i = 0; i < mu; ++i)
+        weights[i] /= s1;
+
+      if (mu < 1 || mu > lambda || (mu == lambda && weights[0] == weights[mu - 1]))
+        throw std::runtime_error("setWeights(): invalid setting of mu or lambda");
+    
+
+    if (cs > 0)
+      cs *= (mueff + 2.) / (N + mueff + 3.);
+    if (cs <= 0 || cs >= 1)
+      cs = (mueff + 2.) / (N + mueff + 3.);
+
+    if (ccumcov <= 0 || ccumcov > 1)
+      ccumcov = 4. / (N + 4);
+
+    if (mucov < 1)
+      mucov = mueff;
+    double t1 = 2. / ((N + 1.4142)*(N + 1.4142));
+    double t2 = (2.* mueff - 1.) / ((N + 2.)*(N + 2.) + mueff);
+    t2 = (t2 > 1) ? 1 : t2;
+    t2 = (1. / mucov)* t1 + (1. - 1. / mucov)* t2;
+    if (ccov >= 0)
+      ccov *= t2;
+    if (ccov < 0 || ccov > 1)
+      ccov = t2;
+
+    if (diagonalCov < 0)
+      diagonalCov = 2 + 100. * N / sqrt((double) lambda);
+
+    if (stopMaxFunEvals <= 0)
+      stopMaxFunEvals = facmaxeval * 900 * (N + 3)*(N + 3);
+    else
+      stopMaxFunEvals *= facmaxeval;
+
+    if (stopMaxIter <= 0)
+      stopMaxIter = ceil((double) (stopMaxFunEvals / lambda));
+
+    if (damps < double(0))
+      damps = double(1);
+    damps = damps
+        * (double(1) + double(2)*std::max(double(0), std::sqrt((mueff - double(1)) / (N + double(1))) - double(1)))
+        * (double) std::max(double(0.3), double(1) - // modify for short runs
+          (double) N / (double(1e-6) + std::min(stopMaxIter, stopMaxFunEvals / lambda)))
+        + cs;
+
+    if (updateCmode.modulo < 0)
+      updateCmode.modulo = 1. / ccov / (double) N / 10.;
+    updateCmode.modulo *= facupdateCmode;
+    if (updateCmode.maxtime < 0)
+      updateCmode.maxtime = 0.20; // maximal 20% of CPU-time
+  
+  }
+
+  template<typename funcType>
+  double CMAES<funcType>::Optimize(arma::mat& arr)
+  {
+    arFunvals.set_size(lambda);
+    init(arFunvals);
+
+  while(!testForTermination())
+  {
+    // Generate lambda new search points, sample population
+    samplePopulation();
+
+    arma::mat fit(1,N);
```

Comment:
We should declare ``fit``outside the while loop and reuse the fit instance in each iteration. What do you think?